IPv6 Client Socket Support

dantpm opened the pull request:
Add IPv6 client socket support.
This results in dual-stack IPv4/IPv6 support for client socket connections.

Fixes #876.

Signed-off-by: Dan Anderson <daniel.anderson@intel.com>

AndreasFuchsSIT commented on src/util/io.c:
```diff
@@ -140,33 +143,55 @@ socket_close (
     return TSS2_RC_SUCCESS;
 }
 
+static void uinttostring(unsigned int num, char *str) {
```
use itoa() or snprintf(), but don't reinvent the wheel please...

dantpm replied:
OK, although generally libraries should not report errors, including using fprintf(). That's up to the application. But I see the rule is broken already for tpm2-tss, so I replaced it with snprintf().

Also itoa() does not exist in libc (although atoi does). Function itoa may exist in other libraries, but it is nonstandard.

AndreasFuchsSIT commented on src/util/io.c:
```diff
@@ -140,33 +143,55 @@ socket_close (
 {
-    struct sockaddr_in sockaddr = { 0 };
+    static const struct addrinfo hints = { .ai_socktype = SOCK_STREAM,
+        .ai_protocol = IPPROTO_TCP };
```
Just for clearity, please set .ai_family = AF_UNSPEC

dantpm replied:
OK. That's the default (0), as with all fields, but it does make it explicit.

AndreasFuchsSIT commented on src/util/io.c:
```diff
@@ -140,33 +143,55 @@ socket_close (
-    ret = connect (*sock, (struct sockaddr*)&sockaddr, sizeof (sockaddr));
-    if (ret == -1) {
+    
+    LOG_DEBUG ("Connecting TCP stream socket to host %s on port %" PRIu16,
```
Please put some the LOG_DEBUG with the actual target node inside the loop instead. I think that gives way more information...

If you set the hints.ai_flags = AI_CANONIDN then p->ai_canonname should also be set correctly. Never used it myself, but should work...

dantpm replied:
I added it, but, although it is useful, be warned that AI_CANONIDN is a non-standard glib-specific extension.

dantpm replied:
OK. Because AI_CANONIDN is non-standard. It is a glib-specific extension, I had to add a #define __USE_GNU to break the standards enforcement in netdb.h.

AndreasFuchsSIT replied:
Well the actual point was to bring this LOG-Statement inside the `for`-loop, so we see which IPs the lib is attempting to connect to...
I guess a __USE_GNU addition would not be the nicest...
Google gives us an alternative: `inet_ntop` see below...

tstruk commented on src/util/io.c:
```diff
@@ -146,27 +152,41 @@ socket_connect (
+    struct addrinfo *retp = NULL;
+    struct addrinfo *pprev = NULL;
+    struct addrinfo *p;
+    char port_str[MAX_PORT_STR_LEN];
```
char port_str[MAX_PORT_STR_LEN + 1]; sizeof doesn't count the terminating NULL.

dantpm replied:
Yes it does count the terminating NUL character.  You may be confusing sizeof() with strlen().  Here's an example:

#include <stdio.h>
#include <string.h>
void main() {
        printf("strlen of \"foo\" is %lu\n", strlen("foo"));
        printf("sizeof of \"foo\" is %lu\n", sizeof("foo"));
}
(glacier ~/tmp) ./foo
strlen of "foo" is 3
sizeof of "foo" is 4

AndreasFuchsSIT commented on src/util/io.c:
```diff
@@ -146,31 +152,48 @@ socket_connect (
+    
+    LOG_DEBUG ("Connecting TCP stream socket to host %s on port %" PRIu16,
+               hostname, port);
+    for (p = retp; p != NULL; pprev = p, p = p->ai_next) {
```
FWIW, Somewhere at the top with attribute unused because of MAXLOGLEVEL=NONE:
```
char buff[255] __attribute__((unused));
```
Here:
```
LOG_DEBUG("Connecting to %s", inet_ntop(p->ai_family, p->ai_addr, &buff[0], sizeof(buff)));
```

dantpm replied:
I added this, inside the for() loop.
Connection attempt failures to multi-homed hosts are not an error, but since this is just a debug message it's OK to be in the loop.

AndreasFuchsSIT commented on src/util/io.c:
```diff
@@ -146,31 +152,48 @@ socket_connect (
-        socket_close (sock);
+                     "%d: %s", (pprev != NULL && pprev->ai_canonname != NULL) ?
+                      pprev->ai_canonname : hostname,
+                      port, errno, strerror (errno));
```
For this purpose, I'd say that the flag `AI_CANONNAME` and printing `retp->ai_canonname` are more meaningful.

Of course only before the freeaddrinfo, as it should have been here anyways....

dantpm commented:
I moved the LOG_DEBUG message inside the loop to print every hostname connection attempted.

I also removed the GNU-specific, non-standard  AI_CANONNAME and ai_canonname  and used inet_ntop() as suggested.

dantpm commented:
Rebased with upstream and addressed all comments.

AndreasFuchsSIT commented:
@dantpm after #910 this is now conflictig... Sorry, you'll have to rebase again...

dantpm commented:
I just rebased yet again. It has the changes from #910.

dantpm commented:
@tstruk, Philip just mentioned that @AndreasFuchsSIT is on vacation.  The last message he left is that I needed to do another rebase before integration, but never did a pull.
All the issues are addressed.

Could you look at it and do a pull?

tstruk commented:
@dantpm merged. One minor comment. In io.c line 192 comment uses C++ style:
+            break; // socket connected OK
Please use C style comments.
I will update this one.

dantpm commented:
I fixed it and rebased again.
